### ISSUE TITLE
SPU2-X Portaudio error: Invalid device on some machines

### ISSUE THREAD

**Issue by ramapcsx2** [PCSX2/pcsx2#390]
It seems the default device selection can fail / choose an unfit device on some configurations, like my Windows 8 laptop with some HD Audio device.

**Comment by karasuhebi** [comment]
I also have this issue with PortAudio on my Windows 8.1 machine. I have more than one valid audio output device and I suspect this has something to do with it. Here's the info I've gathered so far:

With Host Api set to Unspecified, I get sound on my speakers. But if I change my Default Device via Windows to my monitor's speakers (since I'm not given any other choice for devices in the SPU2-X GUI), I get an error: * SPU2-X: PortAudio error: Invalid device. UPDATE: When I changed back my default device to my speakers (via Windows), Unspecified was no longer providing me with sound. Weird.

With Host Api set to DirectSound I get some weird buzzing sound after a couple of seconds. Doesn't matter what device I choose.

With Host Api set to WASAPI and Device Name set to my speakers (Realtek High Definition Audio) it won't work; I get the "* SPU2-X: PortAudio error: Invalid device" error. But if I select 'WASAPI Exclusive Mode' without changing any other setting, I get no such error and I do in fact get sound from my speakers. Strange.

With Host Api set to WASAPI and Device Name set to my monitor's speakers (AMD High Definition Audio Device), I do get sound.

With Host Api set to WDM-KS I get sound, doesn't matter which of the two devices I pick. Weirdly enough though (and quite possibly unrelated), if I leave Device Name on 'Default Device', it outputs sound from my monitor, not my speakers. Strange. I guess Windows must have a setting somewhere other than the Control Panel's Sound window that stores a different value for Default Device. KS stands for Kernel Streaming so maybe it's a kernel setting? Who knows.

In conclusion, I cannot get sound to my speakers via any other Host API that's not WDM-KS or WASAPI Exclusive Mode, which I don't want to use. This is why I use XAudio 2 instead of PortAudio.

P.S. I'm not sure why SPU2-X's author decided to go with 'Host Api' instead of 'Host API' in the GUI. Same thing with why he chose to go with Portaudio instead of the correctly-cased PortAudio. Even the log prints it as PortAudio...you can see in the error I included above.

**Comment by TheLastRar** [comment]
I encountered this issue when using WASAPI (non-exclusive). (despite a device being explicitly selected).

I discovered that audio format the device format was set to was "24bit 192000 Hz (Studio Quality)". Setting this to "24bit 4800 Hz (Studio Quality)" allowed SPU2-X to use the device with WASAPI (non-exclusive).

might this hint at what the issue is on other systems.

**Comment by MrCK1** [comment]
Did some testing on this, the only mode that works for me is WDM-KS. DirectSound gives me computer noise, and I can't get any output from WASAPI (console reports invalid device)

The invalid device error message seems to be related to exclusive mode somehow. [This ticket](https://app.assembla.com/spaces/portaudio/tickets/147-possible-issue-with-wasapi-exclusive-mode/details#) was opened in PortAudio back in 2011, but closed because it couldn't be reproduced apparently. Whatever this issue is, it's a problem with PortAudio.

**Comment by mirh** [comment]
Portaudio works for me. 
The only new thing I have to report is that WASAPI works too now (guess like that might have been fixed by upgrade to svn1954)
Oh, and [asio](https://github.com/PCSX2/pcsx2/issues/2024#issuecomment-324311335). ^^

p.s. of course I'm still on windows 7

**Comment by MrCK1** [comment]
@mirh So everything works fine for you on a newer build of PortAudio?

**Comment by mirh** [comment]
An update has been pushed since my last report above dated 2014, that's what I meant. 
Atm everything works fine for me, but directsound with latencies under 19ms (including default "smallest")

**Comment by MrCK1** [comment]
@mirh and anybody else. Try this SPU2-X build with the newest version of PortAudio. For whatever reason, I can't get audio unless I set Host Api to "Unspecified" 

See if any behavior has changed for you guys.

Rename to .dll [Spu2-X.pdf](https://github.com/PCSX2/pcsx2/files/1347955/Spu2-X.pdf)

**Comment by mirh** [comment]
Very, very funny, since for me this plugin seemed to have broken "Unspecified", which isn't working anymore. 

In other news, I just realized I had never reported about WDM-KS, which is not working in both normal dll and yours.

**Comment by MrCK1** [comment]
I never bothered to re-test PortAudio since I got a dedicated soundcard a few weeks ago. Basically, the "Invalid Device" error is a Realtek specific issue it seems. 

My SoundBlaster X-Fi Titanium works in all modes, with no errors shown in the console. The only module I had issues with was DirectSound - which produces robotic noise.

**Comment by atomic83GitHub** [comment]
Yes, I tested SPU2-x on a realtek and a IDT audio device, booth fail to load it, but a dedicated audio device (like the terratec aureon 7.1 usb) work as @MrCK1 say.

**Comment by mirh** [comment]
*looks at my results nobody seems to have noticed*
Had no problems with both Realtek and Xonar (also, mandatory fuck you creative)

**Comment by rcaridade145** [comment]
This revison 
https://app.assembla.com/spaces/portaudio/git/commits/bc40f42cba713db3349bb1120b5d3475dd45e37e

states 

> wasapi: use IAudioClient3 for Windows 10 for best compatibility 

Maybe it can help

**Comment by MrCK1** [comment]
Unfortunately, we don't use development builds.of PortAudio in PCSX2 and I have a PR to merge the next stable version in anyways.

The remaining issues are down to integrated audio and crappy drivers - IIRC it has nothing to do with PortAudio either. There's not really much we can do about it on our side.

**Comment by rcaridade145** [comment]
> Unfortunately, we don't use development builds.of PortAudio in PCSX2 and I have a PR to merge the next stable version in anyways.

I only mentioned that revision because the ticket to merge the dev build of PortAudio already has 2 extra revisions

> 2 commits that were pushed after release have been included here as well. (A typo fix for CMakeLists.txt and another small patch).

**Comment by mirh** [comment]
> The remaining issues are down to integrated audio and crappy drivers 

I don't know, what you say seems kinda sound - but then I wonder why I, for one, am doing totally fine with my realtek. On W7.

**Comment by MrCK1** [comment]
Hmm, I guess you're right. Everything works on my laptop's Realtek chip - only caveat is that WASAPI requires exclusive mode enabled. The new plugin from #1912 behaves the same as well, so I guess Windows 10's issues may have been ironed out?

Can I have some others do re-tests on their systems with the current and new plugin from #1912/ report back with any issues? If not, hopefully we can get it merged soon. ;)

**Comment by ramapcsx2** [comment]
No issues with using the #1912 linked .dll. Thorough testing is difficult, of course.
I tend to hope that never versions of libraries iron out bugs, so if they build and pass some tests, I'd merge the updates.

**Comment by mirh** [comment]
> only caveat is that WASAPI requires exclusive mode enabled

That ain't even required. 
The only thing, again, I have to report (aside of WDM-KS) is that direcstound requires a latency higher than a certain X

**Comment by MrCK1** [comment]
So I guess this issue is resolved then? Any objections on closing this?

**Comment by mirh** [comment]
TBH, considering all the variety of our findings, I think this has always been more about individual hw/sw configs than an actual problem with pcsx2. 

Said this, I wonder why latest portaudio still hasn't been merged.

**Comment by turtleli** [comment]
> Unfortunately, we don't use development builds.of PortAudio in PCSX2

Current version in master is a dev build...
